File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/ui/screens/TransferServerScreens.kt
// Transfer-a-box Compose screens (Layer C, Android). Mirror of iOS
// FlagshipUI/Screens/TransferServerScreens.swift + the webapp giver/acquirer UX.

package com.flagshipserver.app.ui.screens

import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.flagshipserver.app.core.qrImageBitmap
import com.flagshipserver.app.ui.components.FSDangerButton
import com.flagshipserver.app.ui.components.FSField
import com.flagshipserver.app.ui.components.FSPrimaryButton
import com.flagshipserver.app.ui.components.FSSecondaryButton
import com.flagshipserver.app.ui.theme.FS
import com.flagshipserver.app.viewmodels.TransferAcquirerPhase
import com.flagshipserver.app.viewmodels.TransferAcquirerViewModel
import com.flagshipserver.app.viewmodels.TransferGiverPhase
import com.flagshipserver.app.viewmodels.TransferGiverViewModel
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

@Composable
private fun TransferCallout(text: String, color: androidx.compose.ui.graphics.Color) {
    Text(
        text,
        color = FS.colors.text,
        style = TextStyle(fontSize = 13.sp, lineHeight = 18.sp),
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(FS.radius.sm))
            .padding(FS.space.s3),
    )
}

/** GIVER: irreversible "Transfer to another account" → type-confirm + biometric
 *  → QR. */
@Composable
fun TransferGiverScreen(vm: TransferGiverViewModel, serverDomain: String) {
    val phase by vm.phase.collectAsState()
    val scope = rememberCoroutineScopeForTransfer()
    var typed by remember { mutableStateOf("") }
    val confirmed = typed.lowercase() == serverDomain.lowercase()
    val scroll = rememberScrollState()

    Column(
        Modifier.fillMaxSize().verticalScroll(scroll).padding(FS.space.s6),
        verticalArrangement = Arrangement.spacedBy(FS.space.s4),
    ) {
        when (val p = phase) {
            is TransferGiverPhase.AwaitingClaim, is TransferGiverPhase.Resealing -> {
                Text("Have the new owner scan this", color = FS.colors.text, style = TextStyle(fontSize = 18.sp, fontWeight = FontWeight.SemiBold))
                Text(
                    "On their phone: Add a server → Take over a transferred box. Keep this screen open until it completes — your phone hands off the disk key after they claim it.",
                    color = FS.colors.textMuted, style = TextStyle(fontSize = 13.sp, lineHeight = 18.sp),
                )
                vm.qrText?.let { Image(qrImageBitmap(it), contentDescription = "Transfer QR", modifier = Modifier.size(240.dp)) }
                if (p is TransferGiverPhase.Resealing) {
                    TransferCallout("They claimed it — handing off the disk key…", FS.colors.primary)
                } else {
                    Text("Waiting for the new owner…", color = FS.colors.textMuted)
                }
                LaunchedEffect(Unit) {
                    while (vm.phase.value is TransferGiverPhase.AwaitingClaim) {
                        if (vm.pollOnce()) break
                        delay(3000)
                    }
                }
            }
            is TransferGiverPhase.Completed -> {
                TransferCallout("Transfer complete. $serverDomain now belongs to the new owner${p.newServerDomain?.let { " (now $it)" } ?: ""}.", FS.colors.success)
                Text("It's no longer in your fleet.", color = FS.colors.textMuted)
            }
            else -> {
                if (p is TransferGiverPhase.Failed) TransferCallout(p.message, FS.colors.danger)
                Text("Transfer to another account", color = FS.colors.text, style = TextStyle(fontSize = 24.sp, fontWeight = FontWeight.Medium))
                Text(
                    "This hands $serverDomain and ALL its contents to another account. You will lose control of it. This cannot be undone.",
                    color = FS.colors.textMuted, style = TextStyle(fontSize = 14.sp, lineHeight = 20.sp),
                )
                FSField(value = typed, onValueChange = { typed = it }, label = "Type the server's address to confirm", placeholder = serverDomain, fieldTag = "transfer-confirm-field")
                val working = p is TransferGiverPhase.Signing || p is TransferGiverPhase.Posting
                FSDangerButton(
                    label = if (working) "Working…" else "Transfer this box",
                    onClick = { if (confirmed && !working) scope.launch { vm.start() } },
                    enabled = confirmed && !working,
                    block = true,
                )
            }
        }
    }
}

/** ACQUIRER: "Take over a transferred box" — scan the giver's QR, confirm to claim. */
@Composable
fun TransferAcquirerScreen(vm: TransferAcquirerViewModel) {
    val phase by vm.phase.collectAsState()
    val scope = rememberCoroutineScopeForTransfer()
    var scanning by remember { mutableStateOf(true) }
    val scroll = rememberScrollState()

    Column(
        Modifier.fillMaxSize().verticalScroll(scroll).padding(FS.space.s6),
        verticalArrangement = Arrangement.spacedBy(FS.space.s4),
    ) {
        when (val p = phase) {
            is TransferAcquirerPhase.Idle -> {
                Text("Point your camera at the transfer code", color = FS.colors.text, style = TextStyle(fontSize = 18.sp, fontWeight = FontWeight.SemiBold))
                Text("The current owner shows it from their box's page (Transfer to another account).", color = FS.colors.textMuted, style = TextStyle(fontSize = 13.sp, lineHeight = 18.sp))
                Column(Modifier.fillMaxWidth().height(320.dp).clip(RoundedCornerShape(FS.radius.md))) {
                    QRScanner(onScanned = { text -> if (scanning) { scanning = false; vm.ingest(text) } })
                }
            }
            is TransferAcquirerPhase.Scanned -> {
                Text("Take over this box?", color = FS.colors.text, style = TextStyle(fontSize = 24.sp, fontWeight = FontWeight.Medium))
                Text("You'll become the owner of ${p.serverDomain} and all its contents. The current owner loses control of it.", color = FS.colors.textMuted, style = TextStyle(fontSize = 14.sp, lineHeight = 20.sp))
                FSPrimaryButton(label = "Take ownership", onClick = { scope.launch { vm.confirm() } }, block = true)
            }
            is TransferAcquirerPhase.Signing, is TransferAcquirerPhase.Posting -> {
                CircularProgressIndicator()
                Text("Claiming the box…", color = FS.colors.textMuted)
            }
            is TransferAcquirerPhase.Claimed -> {
                TransferCallout("You now own ${p.newServerDomain ?: "this box"}. It will come online under your account shortly.", FS.colors.success)
            }
            is TransferAcquirerPhase.Failed -> {
                TransferCallout(p.message, FS.colors.danger)
                FSSecondaryButton(label = "Scan again", onClick = { vm.resetForRescan(); scanning = true })
            }
        }
    }
}

@Composable
private fun rememberCoroutineScopeForTransfer() = androidx.compose.runtime.rememberCoroutineScope()

File: apps/mobile/ios/Sources/FlagshipUI/Screens/TransferServerScreens.swift
import SwiftUI
import FlagshipAPI
import FlagshipCore

/// Small inline callout (success/info/danger) — the transfer screens don't need
/// the full toast machinery, just a tinted strip.
private struct TransferCallout: View {
    @Environment(\.colorScheme) private var scheme
    enum Kind { case success, info, danger }
    let kind: Kind
    let text: String
    var body: some View {
        let c = FSColors.scheme(scheme)
        let color: Color = kind == .success ? c.success : (kind == .danger ? c.danger : c.primary)
        Text(text)
            .font(FS.font.bodySm())
            .foregroundColor(c.text)
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(FS.space.s3)
            .background(color.opacity(0.12))
            .clipShape(RoundedRectangle(cornerRadius: FS.radius.sm))
    }
}

/// GIVER: the irreversible "Transfer to another account" flow on server-detail.
/// Type-to-confirm the FQDN, pass the biometric, then show the QR for the
/// acquirer to scan. Mirrors the webapp giver card.
public struct TransferGiverScreen: View {
    @Environment(\.colorScheme) private var scheme
    @State private var vm: TransferGiverViewModel
    @State private var typed = ""
    private let serverDomain: String

    public init(vm: TransferGiverViewModel, serverDomain: String) {
        _vm = State(initialValue: vm)
        self.serverDomain = serverDomain
    }

    private var confirmed: Bool { typed.lowercased() == serverDomain.lowercased() }

    public var body: some View {
        let c = FSColors.scheme(scheme)
        ScrollView {
            VStack(alignment: .leading, spacing: FS.space.s4) {
                switch vm.phase {
                case .idle, .signing, .posting:
                    warning(c)
                case .awaitingClaim, .resealing:
                    qrSection(c)
                case .completed(let newDomain):
                    completed(c, newDomain)
                case .failed(let msg):
                    TransferCallout(kind: .danger, text: msg)
                    warning(c)
                }
            }
            .padding(FS.space.s6)
        }
        .background(c.bg.ignoresSafeArea())
        .navigationTitle("Transfer box")
        .navigationBarTitleDisplayMode(.inline)
    }

    private var working: Bool {
        if case .signing = vm.phase { return true }
        if case .posting = vm.phase { return true }
        return false
    }

    @ViewBuilder private func warning(_ c: FSColors) -> some View {
        VStack(alignment: .leading, spacing: FS.space.s3) {
            Text("Transfer to another account")
                .font(FS.font.h2()).foregroundColor(c.text)
            Text("This hands \(serverDomain) and ALL its contents to another account. You will lose control of it. This cannot be undone.")
                .font(FS.font.body()).foregroundColor(c.textMuted)
            Text("Type the server's address to confirm:")
                .font(FS.font.bodySm()).foregroundColor(c.textMuted)
            TextField(serverDomain, text: $typed)
                .textInputAutocapitalization(.never)
                .autocorrectionDisabled()
                .padding(FS.space.s3)
                .background(c.surface)
                .clipShape(RoundedRectangle(cornerRadius: FS.radius.sm))
                .accessibilityIdentifier("transfer-confirm-field")
            FSDangerButton(working ? "Working…" : "Transfer this box", block: true) {
                if confirmed && !working { Task { await vm.start() } }
            }
            .opacity(confirmed ? 1 : 0.4)
            .accessibilityIdentifier("transfer-start")
        }
    }

    @ViewBuilder private func qrSection(_ c: FSColors) -> some View {
        VStack(alignment: .leading, spacing: FS.space.s4) {
            Text("Have the new owner scan this")
                .font(FS.font.h3()).foregroundColor(c.text)
            Text("On their phone: Add a server → Take over a transferred box. Keep this screen open until it completes — your phone hands off the disk key after they claim it.")
                .font(FS.font.bodySm()).foregroundColor(c.textMuted)
            if let text = vm.qrText {
                HStack { Spacer(); PairingQRView(text: text, size: 240); Spacer() }
            }
            if case .resealing = vm.phase {
                TransferCallout(kind: .info, text: "They claimed it — handing off the disk key…")
            } else {
                ProgressView("Waiting for the new owner…")
            }
        }
        .task {
            while case .awaitingClaim = vm.phase {
                if await vm.pollOnce() { break }
                try? await Task.sleep(nanoseconds: 3_000_000_000)
            }
        }
    }

    @ViewBuilder private func completed(_ c: FSColors, _ newDomain: String?) -> some View {
        VStack(alignment: .leading, spacing: FS.space.s3) {
            TransferCallout(kind: .success, text: "Transfer complete. \(serverDomain) now belongs to the new owner\(newDomain.map { " (now \($0))" } ?? "").")
            Text("It's no longer in your fleet.").font(FS.font.body()).foregroundColor(c.textMuted)
        }
    }
}

/// ACQUIRER: "Take over a transferred box" — scan the giver's QR, then confirm
/// to claim ownership. Mirrors the webapp acquirer entry on the add-server
/// chooser.
public struct TransferAcquirerScreen: View {
    @Environment(\.colorScheme) private var scheme
    @State private var vm: TransferAcquirerViewModel
    @State private var scanning = true

    public init(vm: TransferAcquirerViewModel) {
        _vm = State(initialValue: vm)
    }

    public var body: some View {
        let c = FSColors.scheme(scheme)
        ScrollView {
            VStack(alignment: .leading, spacing: FS.space.s4) {
                switch vm.phase {
                case .idle:
                    scanner(c)
                case .scanned(let domain):
                    confirm(c, domain)
                case .signing, .posting:
                    ProgressView("Claiming the box…")
                case .claimed(let newDomain):
                    TransferCallout(kind: .success, text: "You now own \(newDomain ?? "this box"). It will come online under your account shortly.")
                case .failed(let msg):
                    TransferCallout(kind: .danger, text: msg)
                    FSSecondaryButton("Scan again") { vm.resetForRescan(); scanning = true }
                }
            }
            .padding(FS.space.s6)
        }
        .background(c.bg.ignoresSafeArea())
        .navigationTitle("Take over a box")
        .navigationBarTitleDisplayMode(.inline)
    }

    @ViewBuilder private func scanner(_ c: FSColors) -> some View {
        VStack(alignment: .leading, spacing: FS.space.s4) {
            Text("Point your camera at the transfer code")
                .font(FS.font.h3()).foregroundColor(c.text)
            Text("The current owner shows it from their box's page (Transfer to another account).")
                .font(FS.font.bodySm()).foregroundColor(c.textMuted)
            QRScannerView(
                onScan: { text in if scanning { scanning = false; _ = vm.ingest(text) } },
                validate: { ServerTransferFlow.looksLikeTransferQR($0) }
            )
            .frame(height: 320)
            .clipShape(RoundedRectangle(cornerRadius: FS.radius.md))
            .accessibilityIdentifier("transfer-scanner")
        }
    }

    private var claiming: Bool {
        if case .signing = vm.phase { return true }
        if case .posting = vm.phase { return true }
        return false
    }

    @ViewBuilder private func confirm(_ c: FSColors, _ domain: String) -> some View {
        // Take-over is IRREVERSIBLE (ownership change + disk re-seal) ⇒ a SEVERE
        // tiered confirm: danger color + type-the-FQDN + the claim biometric
        // (fired inside vm.confirm()). What you see (domain) is what you sign —
        // the same verified offer bytes drive both the display and the claim.
        VStack(alignment: .leading, spacing: FS.space.s4) {
            Text("Take over this box?").font(FS.font.h2()).foregroundColor(c.text)
            TieredConfirmationSheet(
                severity: .severe,
                title: "You will become the owner of \(domain)",
                message: "You'll own \(domain) and everything on it. The current owner loses control of it. This cannot be undone.",
                confirmPhrase: domain,
                confirmFieldLabel: "Type the box's address to confirm:",
                actionTitle: "Take ownership",
                busy: claiming
            ) {
                Task { await vm.confirm() }
            }
            .accessibilityIdentifier("transfer-claim")
        }
    }
}
